In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/titanic-dataset-1/train.csv
/kaggle/input/titanic-dataset-1/test.csv
/kaggle/input/titanic-dataset-1/gender_submission.csv


In [2]:
import numpy as np 
import pandas as pd 
import os

In [3]:
# check the work directory 
os.getcwd()

'/kaggle/working'

In [4]:
# back to kaggle 
os.chdir('/kaggle/') 
os.getcwd()

'/kaggle'

In [5]:
# pass the file path to os.listdir() function to list all of the objects 
os.listdir('/kaggle') 

# to see what in the input:
os.listdir('/kaggle/input')  

os.listdir('/kaggle/input/titanic-dataset-1')  

['train.csv', 'test.csv', 'gender_submission.csv']

In [6]:
# supply the file path 
titanic_train = pd.read_csv('input/titanic-dataset-1/train.csv') 

In [7]:
# It's a good idea to start off by checking the dimensions of your data set with df.shape 
# and the variable data types of df.dtypes.

titanic_train.shape 

# 891 records and 12 columns 

(891, 12)

In [8]:
titanic_train.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [9]:
titanic_train.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [10]:
# check the first 5 rows 

titanic_train.head(5) 

# a mixture of numeric columns and columns with text data.  


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [11]:
# look at a statistical summary of the variables with df.describe(): 

# describe() function can not summarize categorical columns

titanic_train.describe()  

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Describe the categorical variables: 

In [12]:
# make a list of categorical columns 

# step1: find the cols where dtype is an object  
# code: titanic_train.dtypes == 'object' 

# step2: take the index of those cols: 
categorical = titanic_train.dtypes[titanic_train.dtypes == 'object'].index 


# use 'categorical' to index back into titanic_train 
# rerun the describe():   
titanic_train[categorical].describe()  
 

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

In [13]:
# VARIABLE DESCRIPTIONS:
# survival        Survival
#                 (0 = No; 1 = Yes)
# pclass          Passenger Class
#                 (1 = 1st; 2 = 2nd; 3 = 3rd)
# name            Name
# sex             Sex
# age             Age
# sibsp           Number of Siblings/Spouses Aboard
# parch           Number of Parents/Children Aboard
# ticket          Ticket Number
# fare            Passenger Fare
# cabin           Cabin
# embarked        Port of Embarkation
#                 (C = Cherbourg; Q = Queenstown; S = Southampton) 

In [14]:
# After looking at the var for the first time, you should ask yourself a few questions, 
# and solve these questions later: 
# 1. Do I need all of the variables?
# 2. Should I transform any variables?
# 3. Are there NA values, outliers or other strange values?
# 4. Should I create new variables?    


This data set is provided in conjunction with a predictive modeling competition where the goal is to use the training data to predict whether passengers of the titanic listed in a second data set survived or not. We won't be dealing with the second data set (known the test set) right now, but we will revisit this competition and make predictions in a future lesson on predictive modeling. 

# Do I need all of the variables? 

In [15]:
# DATA CLEANING 

In [16]:
titanic_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [17]:
# "PassengerId" is just a number assigned to each passenger. 
# It is nothing more than an arbitrary identifier;
# Let's remove it anyway.   

del titanic_train['PassengerId'] 


1. "Survived" indicates whether each passenger lived or died. 
Since predicting survival is our goal, we definitely need to keep it.

2. Features that describe passengers numerically or group them into a few broad categories could be useful for predicting survival.

   
The variables Pclass, Sex, Age, SibSp, Parch, Fare, and Embarked appear to fit this description, so let's keep all of them. 



In [18]:
# "Name" appears to be a character string of the name of each passenger. 

In [19]:
#  This command sorts only the first 15 names in the 'Name' column: 
#  Wrong Version  

sorted(titanic_train['Name'][0:15])   

['Allen, Mr. William Henry',
 'Andersson, Mr. Anders Johan',
 'Bonnell, Miss. Elizabeth',
 'Braund, Mr. Owen Harris',
 'Cumings, Mrs. John Bradley (Florence Briggs Thayer)',
 'Futrelle, Mrs. Jacques Heath (Lily May Peel)',
 'Heikkinen, Miss. Laina',
 'Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)',
 'McCarthy, Mr. Timothy J',
 'Moran, Mr. James',
 'Nasser, Mrs. Nicholas (Adele Achem)',
 'Palsson, Master. Gosta Leonard',
 'Sandstrom, Miss. Marguerite Rut',
 'Saundercock, Mr. William Henry',
 'Vestrom, Miss. Hulda Amanda Adolfina']

In [20]:
# The correct version:  
# This command sorts the entire 'Name' column

sorted(titanic_train['Name'])[0:15] 

['Abbing, Mr. Anthony',
 'Abbott, Mr. Rossmore Edward',
 'Abbott, Mrs. Stanton (Rosa Hunt)',
 'Abelson, Mr. Samuel',
 'Abelson, Mrs. Samuel (Hannah Wizosky)',
 'Adahl, Mr. Mauritz Nils Martin',
 'Adams, Mr. John',
 'Ahlin, Mrs. Johan (Johanna Persdotter Larsson)',
 'Aks, Mrs. Sam (Leah Rosen)',
 'Albimona, Mr. Nassef Cassem',
 'Alexander, Mr. William',
 'Alhomaki, Mr. Ilmari Rudolf',
 'Ali, Mr. Ahmed',
 'Ali, Mr. William',
 'Allen, Miss. Elisabeth Walton']

In [21]:
titanic_train['Name'].describe()  

count                         891
unique                        891
top       Braund, Mr. Owen Harris
freq                            1
Name: Name, dtype: object

1. Name variable has 889 unique values.
2. In general, a categorical variable that is unique to each case isn't useful for prediction. 

3. We could extract last names to try to group family members together, but even then the number of categories would be very large.
   
4. On the other hand, it can be nice to have some way to uniquely identify particular cases and names are interesting from a personal and historical perspective, so let's keep Name, knowing that we won't actually use it in any predictive models we make.   

In [22]:
titanic_train['Ticket'][0:15]

0            A/5 21171
1             PC 17599
2     STON/O2. 3101282
3               113803
4               373450
5               330877
6                17463
7               349909
8               347742
9               237736
10             PP 9549
11              113783
12           A/5. 2151
13              347082
14              350406
Name: Ticket, dtype: object

In [23]:
titanic_train['Ticket'].describe()

count        891
unique       681
top       347082
freq           7
Name: Ticket, dtype: object

1. Ticket has 680 unique values: almost as many as there are passengers.
2. Categorical variables with almost as many levels as there are records are often not very useful for prediction.
3. We could try to reduce the number of levels by grouping certain tickets together,
4. But the ticket numbers don't appear to follow any logical pattern we could use for grouping. Let's remove it. 

In [24]:
del titanic_train['Ticket'] 

In [25]:
# Check the first 15 tickets:  
titanic_train['Cabin'][0:15]  

0      NaN
1      C85
2      NaN
3     C123
4      NaN
5      NaN
6      E46
7      NaN
8      NaN
9      NaN
10      G6
11    C103
12     NaN
13     NaN
14     NaN
Name: Cabin, dtype: object

In [26]:
# Wrong code: 
# sorted(titanic_train['Cabin'])[0:15]  

#Python can not compare NaN and C85 which is larger.   



In [27]:
titanic_train['Cabin'].describe() 

count         204
unique        147
top       B96 B98
freq            4
Name: Cabin, dtype: object

1. Cabin has 147 unique values, which indicates it may not be particularly useful for prediction.
2. On the other hand, the names of the levels for the cabin variable seem to have a regular structure: each starts with a capital letter followed by a number.
3. We could use that structure to reduce the number of levels to make categories large enough that they might be useful for prediction.Let's Keep Cabin for now.


1. Removing variables is often more of an art than a science.
   
2. Data projects are iterative processes: you can start with a simple analysis or model using only a few variables and then expand later by adding more and more of the other variables you initially ignored or removed. 

# Should I Transform Any Variables?¶ 

1. When first loading a data set, some of the variables may be encoded as data types that don't fit well with what the data really is or what it means. 

2. For instance, Survived is just an integer variable that takes on the value 0 or 1 depending on whether a passenger died or survived respectively.

3. Variables that indicate a state or the presence or absence of something with the numbers 0 and 1 are sometimes called indicator variables or dummy variables.

4. Indicator variables are essentially just a shorthand for encoding a categorical variable with 2 levels.

5. We could instead encode Survived as a categorical variable with more descriptive categories:

In [28]:
titanic_train['Survived'].describe()

count    891.000000
mean       0.383838
std        0.486592
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: Survived, dtype: float64

In [29]:
new_survived = pd.Categorical(titanic_train['Survived']) 
new_survived = new_survived.rename_categories(['Died', 'Survived'])  

new_survived.describe()  

counts     freqs
categories                  
Died           549  0.616162
Survived       342  0.383838